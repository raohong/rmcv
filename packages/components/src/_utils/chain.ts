const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
