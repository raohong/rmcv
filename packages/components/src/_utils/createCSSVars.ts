import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
