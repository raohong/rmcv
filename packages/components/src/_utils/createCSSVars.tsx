import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
