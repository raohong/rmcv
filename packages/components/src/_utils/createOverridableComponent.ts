import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
