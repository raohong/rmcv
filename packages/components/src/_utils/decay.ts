import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
