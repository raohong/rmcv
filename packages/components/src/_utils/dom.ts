export const isWindow = (win: any): win is Window => win === window;

const getDocRect = () => {
  const html = document.documentElement;

  return {
    left: 0,
    top: 0,
    bottom: html.clientHeight,
    right: html.clientWidth,
    width: html.clientWidth,
    height: html.clientHeight,
  };
};

export type IBCR = {
  left: number;
  top: number;
  bottom: number;
  right: number;
  width: number;
  height: number;
};

export const getScrollOffset = (elem: Window | Element) => {
  if (isWindow(elem)) {
    return {
      scrollLeft: elem.pageXOffset,
      scrollTop: elem.pageYOffset,
    };
  }

  return {
    scrollLeft: elem.scrollLeft,
    scrollTop: elem.scrollTop,
  };
};

export const getScrollSize = (elem: Window | Element) => {
  if (isWindow(elem)) {
    return {
      scrollWidth: document.documentElement.scrollWidth,
      scrollHeight: document.documentElement.scrollHeight,
    };
  }

  return {
    scrollWidth: elem.scrollWidth,
    scrollHeight: elem.scrollHeight,
  };
};

export const getBoundingClientRect = (elem: Window | Element) => {
  if (isWindow(elem)) {
    return getDocRect();
  }

  const rect = elem.getBoundingClientRect();

  return {
    left: rect.left,
    top: rect.top,
    bottom: rect.bottom,
    right: rect.right,
    height: rect.height,
    width: rect.width,
  };
};
