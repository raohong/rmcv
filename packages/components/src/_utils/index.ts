export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
