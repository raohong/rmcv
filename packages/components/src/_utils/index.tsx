export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
