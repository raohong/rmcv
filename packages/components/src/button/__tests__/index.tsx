import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Button from '..';
import { getPrefixCls } from '../../_utils';

test('render correctly', () => {
  const tree = render(<Button />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test(`render with type`, () => {
  render(<Button type="primary" />);
  expect(screen.getByRole('button')).toHaveClass(getPrefixCls(`button-primary`));
});

test(`render with size`, () => {
  render(<Button size="large" />);

  expect(screen.getByRole('button')).toHaveClass(getPrefixCls(`button-size-large`));
});

test('render with block', () => {
  render(<Button block />);

  expect(screen.getByRole('button')).toHaveClass(getPrefixCls('button-block'));
});

test('render with plain', () => {
  render(<Button plain />);

  expect(screen.getByRole('button')).toHaveClass(getPrefixCls('button-plain'));
});

test('render with shape', () => {
  render(<Button shape="round" />);

  expect(screen.getByRole('button')).toHaveClass(getPrefixCls('button-round'));
});

test('render with icon', () => {
  const Arrow = () => <span data-testid="icon" />;

  render(<Button icon={<Arrow />} />);
  expect(screen.getByTestId('icon')).toBeInTheDocument();
});

test('render with disabled', () => {
  const handler = jest.fn();

  render(<Button disabled onClick={handler} />);
  userEvent.click(screen.getByRole('button'));

  expect(handler).not.toHaveBeenCalled();
  expect(screen.getByRole('button')).toHaveClass(getPrefixCls('button-disabled'));
  expect(screen.getByRole('button')).toHaveProperty('disabled');
});

test('render with loading', () => {
  const handler = jest.fn();

  render(<Button loading onClick={handler} />);
  userEvent.click(screen.getByRole('button'));

  expect(handler).not.toHaveBeenCalled();
});

test('render with loadingText', () => {
  const text = 'loading...';
  render(
    <Button loading loadingText={text}>
      <span data-testid="child" />
    </Button>,
  );

  expect(screen.getByRole('button')).toContainHTML(text);
});

test('render with loadingType', () => {
  render(<Button loading loadingType="circular" />);

  expect(screen.getByRole('button').querySelector('svg')).toBeInTheDocument();
});

test('render with loadingSize', () => {
  render(<Button loadingSize={30} loading />);

  expect(
    screen.getByRole('button').querySelector(`.${getPrefixCls('loading-spinner')}`),
  ).toHaveStyle({
    'font-size': '30px',
  });
});
