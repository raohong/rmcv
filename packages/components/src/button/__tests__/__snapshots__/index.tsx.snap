// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <button
    class="rmc-vant-btn rmc-vant-btn-hairline rmc-vant-btn-normal rmc-vant-btn-square"
    role="button"
  />
</DocumentFragment>
`;
